### DISCUSSION TITLE
Pastas Time Series - Sample Up method (potential errors) when solve model

### DISCUSSION THREAD

**Discussion by DataGeoranger** [pastas/pastas#394] [Q&A] [answered]
Hello!

I constructed a pastas model for a monitoring wells and added linear recharge using both rainfall and reference ET data. My observation GW elevation is in unit of feet with frequency of every 10 minute. My rainfall and ET data has unit of inches with frequency  every hour. 

So I processed the data to make sure that my observation data, rainfall, and ET are aligned in unit of feet with same frequency of 10 minute using time series object function in Pastas such as `pseries.update_series(freq="10T", sample_up="divide")`. I confirmed it that all resample results are correct by plotting the time series of the data before sampling and after sampling (screen shots 1 and 2). 

Then during the model solve, I also made sure that my model freq was set to every 10 mins `ml.solve(freq="10T", noise=True)`. However, it seems that Pastas didn't receive the frequency of my observation, rainfall, and ET data correctly when solving, mainly at two places (screen shot 3):

1. Pastas said my observation data contains illegal character even though I forced my frequency to be 10T. Could you please  advise what I could do to identify the illegal character and prevent the warnings? (
2. Pastas didn't receive my resampled rainfall and ET data (freq = 10T) correctly. It firstly tried to sample up both rainfall and ET data using bfill, which I actually picked 'divide'. And then it tried sample down rainfall and ET to freq=D, which is really bizarre since I didn't ask for it. Could you please advise what I missed to prevent Pastas doing it?

I should mention that I haven't add pumping stress yet to match the observation data. I was planning to add stress one by one, but had this error during the process. I also attached my code and dataset in case it would be useful, 
[Diagnose.zip](https://github.com/pastas/pastas/files/8742414/Diagnose.zip)

Thank you very much in advance and look forward to hearing from you!

Best,
Jack

![image](https://user-images.githubusercontent.com/47303462/169581584-9b5663b5-8d3b-4754-92ff-19f86407d131.png)

Screen shot 1(time series plot before sampling)
![image](https://user-images.githubusercontent.com/47303462/169579607-0d93d1a3-debc-4f7c-b489-88820c8a4857.png)

Screen shot 2(time series plot after sampling)
![image](https://user-images.githubusercontent.com/47303462/169579660-b92c2bf4-a75f-4161-8838-7a773fec0bbe.png)

Screen shot 3(errors that doesn't make sense during solving)
![image](https://user-images.githubusercontent.com/47303462/169582403-2b52da43-6438-46f1-af64-76e38020d267.png)
![image](https://user-images.githubusercontent.com/47303462/169581203-a3d41f07-ccb3-4c3f-8f81-7910a48c7529.png)

**Comment by martinvonk** [comment]
Hi @DataGeoranger,

Someone should look at your discussion more thoroughly because I'm not an expert on high-frequency time series analysis. This is also not something that is done regurarly for hydrological time series. However, I can help you with parts of your issue with Pastas:
- For the illegal character please remove the white spaces. 
- To prevent the frequency upsampling and downsampling, specify the frequency when initializing the Model

So to help you quickly use:
`ml = ps.Model(oseries, name=oseries.name.replace(" ", "_"), freq="H")
`
I choose freq="H" here because "10T" does not work. Not sure why.

**Reply by DataGeoranger** [reply]
Thanks! @martinvonk. I was able to fixe illegal character by  remove the white spaces using your code!
However, Pastas still gave me error for the frequency issue, . I particularly specified the frequency when constructed the time series data and also specified freq when solving the model in my original code, and yet Pastas still tried to upsampling and downsampling the time series data that has already been up sampled. That's the mysterious part. Because this is a pumping test data and my oseries were in 10T frequency, so I would like to keep it as 10T, if possible. But thanks again for your tips!
Wonder if @dbrakenhoff if you have any suggestions once you get a chance to look at my original discussion. Thank you!

**Comment by dbrakenhoff** [comment, marked as answer]
Hi Jack,

When you solve a pastas Model, the timeseries are all updated starting from the original timeseries (unless the timeseries does not need to be updated at all). This probably means that the results from your  `pseries.update_series(freq="10T", sample_up="divide")` are not used, instead it is being recalculated starting from the original hourly series according the arguments passed to `ml.solve()`. So `pseries.series_original` is used instead of `pseries.series`. 

The solution is to either update your resampling settings to match what you want, OR to create a new Timeseries using the updated `10T` timeseries you created with `update_series()`.

Then there is the issue with `10T` not being supported. This is a shortcoming in the calculation of the warmup period. Replacing 

https://github.com/pastas/pastas/blob/6104b66abcd516a6da805b22fd444d4eb0fcd75a/pastas/model.py#L91

with 

`            "warmup": 3650 * to_timedelta(freq),`

and adding `to_timedelta` to the imports at the top of the file should resolve this issue.

Putting it all together in code:

```python
# create new pseries with 10T data
pseries.update_series(freq="10T", sample_up="divide")
pseries = ps.TimeSeries(pseries.series, name='precip', settings='prec', 
                        tmin = '2020-10-28 13:00:00', tmax= '2020-11-23 10:00:00')

# ... do same for evap

ml = ps.Model(oseries, freq="10T")  # set model frequency here, otherwise D is used as default
rch = ps.rch.Linear()
rm = ps.RechargeModel(pseries, eseries, recharge=rch, rfunc=ps.Gamma, 
                name="gwrch")
ml.add_stressmodel(rm)
ml.solve()
```
OR

```python

# set settings to match sample_up=divide
psettings = {'sample_up': 'divide',
 'sample_down': 'mean',
 'fill_nan': 0.0,
 'fill_before': 'mean',
 'fill_after': 'mean'}
pseries = ps.TimeSeries(rain, name='precip', settings=psettings, 
                        tmin = '2020-10-28 13:00:00', tmax= '2020-11-23 10:00:00')

# ... do the same for evap

ml = ps.Model(oseries, freq="10T")  # set model frequency here, otherwise D is used as default
rch = ps.rch.Linear()
rm = ps.RechargeModel(pseries, eseries, recharge=rch, rfunc=ps.Gamma, 
                name="gwrch", settings=[psettings, esettings])  # must pass settings explicitly here, else they are ignored and defaults are used
ml.add_stressmodel(rm)
ml.solve()
```

Note that there is still a difference in these two approaches as the initial model parameters are estimated using the original timeseries (so if you put in hourly timeseries, the initial estimates will be based on hourly data). These estimates are done when the RechargeModel is created. 

My recommendation is to first resample the timeseries to your desired timestep and then construct the model. 

Hope that helps!

**Reply by DataGeoranger** [reply]
Thank you again David! Just to confirm if I understand correctly what you suggested below. You meant I should revise the source code under model.py in my local computer by changing line 91 to `"warmup": 3650 * to_timedelta(freq)`. Is this correct?
If yes, should I change the source code back to `"warmup": Timedelta(3650, freq)` when I analyzed model with a frequency of `'D'`or `'H'` etc.

> Then there is the issue with `10T` not being supported. This is a shortcoming in the calculation of the warmup period. Replacing
> 
> https://github.com/pastas/pastas/blob/6104b66abcd516a6da805b22fd444d4eb0fcd75a/pastas/model.py#L91
> 
> with
> 
> ` "warmup": 3650 * to_timedelta(freq),`
> 
> and adding `to_timedelta` to the imports at the top of the file should resolve this issue.

**Reply by dbrakenhoff** [reply]
I believe the new code should work for all cases, so change it in your source code and you should be all set. I'll probably commit the change to pastas dev-branch sometime soon.

**Reply by dbrakenhoff** [reply]
Ah, I was wrong, the code needs to be a bit more complicated:
`(3650 * to_timedelta(freq) if freq[0].isdigit() else Timedelta(3650, freq))`

**Reply by DataGeoranger** [reply]
Thank you David. The change of source code seems work well!

**Reply by dbrakenhoff** [reply]
Hi Jack, can we mark this as answered?

**Reply by DataGeoranger** [reply]
Absolutely. Thanks much for the help! David.
